runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java

```diff
@@ -92,9 +96,58 @@ public void process(ProcessContext c) {
     }
   }
 
+  static class BatchGroupIntoBatchesWithShardedKeyOverrideFactory<K, V>
```

Comment:
So far we don't have plans to do anything fancy for batch but added this override to avoid breaking the transform in batch if `withShardedKey` is specified.

Reply:
Do we need an override or is the default implementation good enough?

Reply:
Two things that made me add the override for `withShardedKeys`
- We insert an explicit GBK for batch stateful dofns (although I feel this should be done in the backend).

https://github.com/apache/beam/blob/562041956efeae2a186b1d815ea6bcd7d54682ae/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/BatchStatefulParDoOverrides.java#L228

- Currently the batch GIB doesn't use keyed states but instead partitions the iterables after GBK:

https://github.com/apache/beam/blob/562041956efeae2a186b1d815ea6bcd7d54682ae/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java#L84

Reply:
Yeah, GBK before batch stateful DoFns should definitely be done on the runner. 

Should we be using the GBK + Iterators.partition on all runners, not just Dataflow?

Reply:
Did you mean that we should use that as a default implementation? I am assuming that GBK + Iterators.partition only works for batch and the default implementation should work for both streaming and batch.

Reply:
I was assuming we could branch on streaming vs. batch.

Reply:
hmm how would the transform know whether the data is bounded or unbounded? I thought in general everything should work without differentiating batch and streaming.

Reply:
Well, we currently give different overrides in the two cases for Dataflow. Everything works, but there are different performance characteristics that encourage different implementations. (One could argue that this is runner-specific logic.) Fine not to change now.